src/sentry/taskworker/consumer_grpc_push.py

```diff
@@ -0,0 +1,92 @@
+"""
+This module is gRPC client that pushes tasks to the taskworker.
+"""
+
+import logging
+import time
+from collections import deque
+from collections.abc import Iterable
+from concurrent.futures import FIRST_COMPLETED, wait
+from concurrent.futures.thread import ThreadPoolExecutor
+
+import grpc
+from sentry_protos.sentry.v1alpha.taskworker_pb2 import (
+    TASK_ACTIVATION_STATUS_PENDING,
+    DispatchRequest,
+    InflightActivation,
+)
+from sentry_protos.sentry.v1alpha.taskworker_pb2_grpc import WorkerServiceStub
+
+from sentry.taskworker.pending_task_store import PendingTaskStore
+
+logger = logging.getLogger("sentry.taskworker.grpc_server")
+
+
+class ConsumerGrpc:
+    def __init__(self, worker_addrs: Iterable[str]) -> None:
+        self.pending_task_store = PendingTaskStore()
+        self.available_stubs = deque(
+            [WorkerServiceStub(grpc.insecure_channel(worker_addr)) for worker_addr in worker_addrs]
+        )
+        self.current_connections = set()
+
+    def start(self):
+        with ThreadPoolExecutor(max_workers=len(self.available_stubs)) as executor:
+            logger.info("Starting consumer grpc with %s threads", len(self.available_stubs))
+            while True:
+                inflight_activation = self._poll_pending_task()
+
+                if len(self.available_stubs) == 0:
+                    done, not_done = wait(self.current_connections, return_when=FIRST_COMPLETED)
+                    self.available_stubs.extend([future.result() for future in done])
+                    self.current_connections = not_done
```

Comment:
I think ideally this should be done via async grpc instead of regular grpc with multithreading, because with async we only perform preemption during IO, where as with multithreading it's up to the discretion of the executor.

But I am not sure how to replicate this behaviour (where we only poll db when a stub is free) with async python, and if the difference is enough to pose measurable performance benefits. So I decided to implement this instead